File: app/src/main/java/com/dialogapp/dialog/util/Event.java
package com.dialogapp.dialog.util;

import androidx.annotation.Nullable;

public class Event<T> {

    private final T content;
    private boolean hasBeenHandled;

    private Event(T content) {
        this.content = content;
    }

    public static <T> Event<T> createEvent(T data) {
        return new Event<>(data);
    }

    public boolean getHasBeenHandled() {
        return hasBeenHandled;
    }

    @Nullable
    public final T getContentIfNotHandled() {
        if (hasBeenHandled)
            return null;
        else {
            hasBeenHandled = true;
            return content;
        }
    }

    public T peekContent() {
        return content;
    }

    private void setHasBeenHandled(boolean hasBeenHandled) {
        this.hasBeenHandled = hasBeenHandled;
    }
}

File: app/src/main/java/com/dialogapp/dialog/util/Event.kt
package com.dialogapp.dialog.util

/**
 * Used as a wrapper for data that is exposed via a LiveData that represents an event.
 */
open class Event<out T>(private val content: T) {

    var consumed = false
        private set // Allow external read but not write

    /**
     * Returns the content and prevents its use again.
     */
    fun consume(): T? {
        return if (consumed) {
            null
        } else {
            consumed = true
            content
        }
    }

    /**
     * Returns the content, even if it's already been handled.
     */
    fun peek(): T = content

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Event<*>

        if (content != other.content) return false
        if (consumed != other.consumed) return false

        return true
    }

    override fun hashCode(): Int {
        var result = content?.hashCode() ?: 0
        result = 31 * result + consumed.hashCode()
        return result
    }
}